article/results.tex

```diff
@@ -56,59 +56,75 @@ \subsection{Best Practice Guidelines}
+Complimentary, and initially possibly appearing contradictory, to the aforementioned advise we recommend to avoid building separate containers for related steps, in particular if supported by the same toolkits, and rather define and use different \emph{entry points} to the same container.
+E.g., a single container with AFNI could be used to access various tools from AFNI.
+Similar approach is adopted by NeuroDesk~\cite{neurodesk} which provides a large collection of entry points for various tools from a smaller set of containers.
 
 \subsubsection{Do not write debug-relevant data inside the container}
 Debug-relevant data, such as intermediary data processing steps or debugging logs should not be deleted by the workflow, and further, should be written to persistent storage.
-When using containers, if this data is written to a hard-coded path, as they would be on a persistent operating system, it will disappear once the container is removed.
-This data is vital for debugging, and thus should not be lost.
+When using some containerization implementations, such as Docker, if file is written to a hard-coded path, as they would be on a persistent operating system, it will disappear once the container is removed.
+Such file might be vital for debugging, and thus should not be lost.
 This can be avoided by making sure that the paths used for intermediary and debugging outputs are bind-mounted to real directories on the parent system, from which they can be freely inspected.
 
 \subsubsection{Parameterize scratch directories}
 Complex workflows commonly generate large amounts of scratch data — intermediary data processing steps, with no other utility than being read by subsequent steps.
 If these data are written to a hard-coded path, multiple executions will lead to race conditions, compromising one or multiple execution attempts.
-This is avoided by parameterizing the path and/or setting a default value based on a unique string (e.g. generated from the timestamps).
+This can be avoided by parameterizing the path and/or setting a default value based on a unique string (e.g. generated from the timestamps).
 When using containers, this should be done at the container initiation level, as the relevant path is the path on the parent system, and not the path inside the container.
 
 \subsubsection{Dependency versions inside container environments should be frozen as soon as feasible}
-The need for image rebuilding also means that assuring functionality in view of frequent updates is a more difficult inside containers than on a persistent system.
-This is compounded by the frequent and API-breaking update schedules of many scientific software packages.
+The need for an entire image rebuilding also means that assuring consistent functionality in view of frequent updates is more difficult inside containers than in an ad-hoc environment.
```

Comment:
I think it's the persistence which is the relevant feature with respect to updating single components, rather than the ad-hoc-ness. I also think ad-hoc is incorrect, it means “for a particular purpose”. In a sense containers are the ad-hoc environments more so than persistent multi-use systems.

Reply:
```
❯ dict "ad hoc"
1 definition found

From WordNet (r) 3.0 (2006) [wn]:

  ad hoc
      adv 1: for one specific case; "they were appointed ad hoc"
      adj 1: often improvised or impromptu; "an ad hoc committee
             meeting"
      2: for or concerned with one specific purpose; "a coordinated
         policy instead of ad hoc decisions"
```
I guess I had more of the "often improved or impromptu" meaning -- that typical environment is just improvised sequence of installation, upgrades etc steps... so I think it fits here.  

What do you mean by persistence here - that container insides do not change?   again - please suggest alternative wording to improve

Reply:
“The need for an entire image rebuilding also means that assuring consistent functionality in view of frequent updates is more difficult inside containers than in persistent environments.”

perhaps even better:

“The need for an entire image rebuilding also means that assuring consistent functionality in view of frequent updates is more difficult inside containers than in interactively managed environments.“

Reply:
I think we are also missing one more aspect... so:
```suggestion
The need for an entire image rebuilding also means that assuring consistent functionality across rebuilds in view of frequent updates is more difficult inside containers than in interactively managed environments.
```
?